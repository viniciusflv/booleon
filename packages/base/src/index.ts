export { View } from './View';
export { Text } from './Text';
