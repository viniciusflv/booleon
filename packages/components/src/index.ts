export * from './Button';
export * from './Field';
