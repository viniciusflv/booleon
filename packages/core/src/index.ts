export * from "./Base";
export * from "./Text";
