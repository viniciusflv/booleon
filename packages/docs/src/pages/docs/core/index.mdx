---
icon: core
title: Booleon Core
description: Booleon core is a collection of functions with a goal to manipulate and compile CSS, based on boolean props.
---

# Introduction

`Booleon` core is a collection of functions with a goal to `manipulate` and `compile` CSS, based on `boolean props`.

# Instalation

```shell
npm i @booleon/core
```

```shell
yarn add @booleon/core
```

# Under the hood

## uniqueClass

Gets an array of classes and returns only unique ones.

```tsx
export const uniqueClass = (...classes: string[]) =>
  Array.from(new Set(classes.filter(Boolean))).join(' ');
```

## stringHash

To create unique classNames it's being used a hash function that uses the string charcodes.
Based on this [article](https://werxltd.com/wp/2010/05/13/javascript-implementation-of-javas-string-hashcode-method/).

```tsx
export function stringHash(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash;
  }
  return String(hash);
}
```

## styleAppender

Append styles if they don't exist.

```tsx
export function styleAppender(className: string, cb: () => string) {
  let styles = document.querySelector(`style[data-booleon="${className}"]`);
  if (!styles) {
    styles = document.createElement('style');
    styles.setAttribute('data-booleon', className);
    document.head.appendChild(styles);
    styles.innerHTML = cb?.();
  }
}
```

## browserPrefixer

For this feature it's been used [style-vendorizer](https://github.com/kripod/style-vendorizer) by
[kripod](https://github.com/kripod), since it's really lightweight,
and on top of it allowing separated or joined attribute and value.

```tsx
export function browserPrefixer(...str: string[]) {
  if (str.map(Boolean).includes(false)) return '';
  if (str.length < 2) {
    const cssPropRegex = /(?<attr>.*):(?<value>.*);/;
    const { attr, value } = cssPropRegex?.exec(str[0])?.groups || {};
    return attr && value ? styleDeclaration(attr, value) : str[0];
  }
  return styleDeclaration(...str);
}
...
function browserPrefixer('position:sticky;');
// 'position:-webkit-sticky;position:sticky;'
```

## cleanTransformFilter

The motivation for this feature was, allowing the [transform]() and
[filter]() modules to have, not only css values, but also its attributes.

PS: BUT for this to work the compiled filters must be near one another.

```tsx
export function cleanTransformFilter(str: string) {
  return [/transform:.*(;transform:)/g, /filter:.*(;filter:)/g].reduce(
    (acc, rgx) =>
      acc.replace(rgx, (match, group) =>
        match.replace(new RegExp(group, 'g'), ''),
      ),
    str,
  );
}
...
cleanTransformFilter('filter:invert(100%);filter:blur(1rem);');
// 'filter:invert(100%);filter:blur(1rem);'
cleanTransformFilter('transform:transform:translateY(-100);transform:translateX(100);');
// transform:translateY(-100)translateX(100);
```

## cssCompiler

The compiler uses the key to find the callback css expression.
The key is `string` or a `symbol`, if is a symbol it'll exec en regular expression to
match key and use its group as the callback value.

```tsx
export function cssCompiler<M extends BooleonModule>(
  key: string,
  value: string | boolean,
  module: M,
) {
  if (!value) return '';
  return browserPrefixer(
    module[key]?.(value) ??
      Object.getOwnPropertySymbols(module).reduce((acc, symbol) => {
        const { regex } =
          /Symbol\((?<regex>.*)\)/g.exec(String(symbol))?.groups ?? {};
        const [, value] = new RegExp(regex).exec(key) ?? [];
        if (value) {
          // @ts-expect-error https://github.com/microsoft/TypeScript/issues/1863
          acc += module[symbol](value);
        }
        return acc;
      }, ''),
  );
}
...
cssCompiler('flex', true, { flex: () => 'display:flex;' });
// 'display:flex;'
cssCompiler('ft_size_30px', true, {
  [Symbol('^ft_size_(.*)')]: (value) => `font-size:${value};`
});
// 'font-size:30px;'
```

## composeProps

The goal of this feature is to get an key value object and prepare it to be used by the
[styleCompiler](#styleCompiler).

```tsx
export function composeProps(props: Props) {
  const recusiveReducer = (
    acc = {},
    splitKey: string[],
    value: Props,
  ): Props => {
    const [key, ...rest] = splitKey;
    if (rest.length) {
      return {
        ...acc,
        [key]: { ...acc[key], ...recusiveReducer(acc[key], rest, value) },
      };
    }
    return { ...acc, css: { ...acc['css'], [key]: value } };
  };

  return Object.keys(props).reduce((acc, key) => {
    if (!props[key]) return acc;
    return { ...acc, ...recusiveReducer(acc, key.split('__'), props[key]) };
  }, {});
}
...
composeProps({ prefix__style: true, style: true, falsy_style: false });
/*
  {
    prefix: {
      css: {
        style: true,
      },
    },
    css: {
      style: true,
    },
  }
*/
```

## prefixerHandlers

It's used by [styleCompiler](#styleCompiler) to wrap de compiled CSS with a prefix match.
Booleon prefix it's a double underline separetad string. Ex:

`md__flex` => `@media(min-width: 1024px){display: flex;}`

```tsx
const media = (breakpoint: string): PrefixHandler => ({
  key,
  value,
  prefixes,
  className,
  recursiveCompiler,
}) => `@media ${breakpoint}{${recursiveCompiler(value)}}`;
```

## defaultPrefixes

List of [prefixerHandlers](#prefixerHandlers) that comes with the lib.

```tsx
export const defaultPrefixes = {
  undefined: () => '',
  css: prefixerHandlers.classes(),
  dark: prefixerHandlers.theme(),
  xs: prefixerHandlers.media('(min-width: 640px)'),
  sm: prefixerHandlers.media('(min-width: 768px)'),
  md: prefixerHandlers.media('(min-width: 1024px)'),
  lg: prefixerHandlers.media('(min-width: 1440px)'),
  xl: prefixerHandlers.media('(min-width: 1920px)'),
  focus: prefixerHandlers.pseudo(':focus'),
  within: prefixerHandlers.pseudo(':focus-within'),
  after: prefixerHandlers.pseudo(':after'),
  before: prefixerHandlers.pseudo(':before'),
  active: prefixerHandlers.pseudo(':active'),
  checked: prefixerHandlers.pseudo(':checked'),
  disabled: prefixerHandlers.pseudo(':disabled'),
  hover: prefixerHandlers.pseudo(':hover'),
  visited: prefixerHandlers.pseudo(':visited'),
  child: prefixerHandlers.pseudo('>*'),
  last: prefixerHandlers.pseudo(' :last-child'),
  first: prefixerHandlers.pseudo(' :first-child'),
  adjacent: prefixerHandlers.pseudo('~*'),
  sibling: prefixerHandlers.pseudo('+*'),
  odd: prefixerHandlers.pseudo(':nth-child(odd)'),
  even: prefixerHandlers.pseudo(':nth-child(even)'),
  kf: prefixerHandlers.keyframe({
    from: '0%',
    quarter: '25%',
    third: '33%',
    half: '50%',
    x2third: '66%',
    x3quarter: '75%',
    to: '100%',
  }),
};
```

## styleCompiler

Uses the props composed by [composeProps](#composeProps) to
[compile CSS](#cssCompiler) applying [prefixes](#defaultPrefixes).

```tsx
export function styleCompiler<
  P extends Props<string, PrefixHandler>,
  M extends BooleonModule
>(
  className: string,
  composedProps: Props,
  booleonModules: M,
  customPrefixes?: P,
) {
  const prefixes = {
    ...customPrefixes,
    ...defaultPrefixes,
  };
  const recursiveCompiler = (props: Props): string => {
    return Object.keys(props).reduce((acc, key) => {
      const value = props[key];
      if (typeof value === 'object') {
        const handler: PrefixHandler =
          prefixes[key] ??
          prefixes[`${Object.keys(prefixes).find((k) => key.startsWith(k))}`];
        return (acc += handler({
          key,
          value,
          prefixes,
          className,
          recursiveCompiler,
        }));
      }
      return (acc += cssCompiler(key, value, booleonModules));
    }, '');
  };

  return cleanTransformFilter(recursiveCompiler(composedProps));
}
```
