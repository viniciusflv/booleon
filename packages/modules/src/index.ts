export * from './constants';
export * from './css';
export * from './types';
