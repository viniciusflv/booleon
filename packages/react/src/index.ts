export * from './hocBooleon';
export * from './useBooleon';
