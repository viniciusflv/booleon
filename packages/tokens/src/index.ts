export default {
  colors: {
    brand: {
      primary: 'hsl(25, 99%, 42%)',
      secondary: 'hsl(25, 100%, 21%)',
    },
    base: {
      emerald: {
        '065': 'hsl(135, 94%, 65%)',
        '040': 'hsl(175, 59%, 40%)',
        '030': 'hsl(68, 100%, 30%)',
        '014': 'hsl(192, 81%, 14%)',
      },
      citrine: {
        '071': 'hsl(31, 100%, 71%)',
        '044': 'hsl(18, 80%, 44%)',
        '042': 'hsl(25, 99%, 42%)',
        '021': 'hsl(25, 100%, 21%)',
      },
      amber: {
        '076': 'hsl(65, 92%, 76%)',
        '035': 'hsl(45, 100%, 35%)',
      },
      sapphire: {
        '077': 'hsl(191, 97%, 77%)',
        '049': 'hsl(205, 69%, 49%)',
      },
      amethyst: {
        '078': 'hsl(265, 89%, 78%)',
        '074': 'hsl(326, 100%, 74%)',
        '051': 'hsl(225, 27%, 51%)',
        '031': 'hsl(232, 14%, 31%)',
        '018': 'hsl(231, 15%, 18%)',
        '007': 'hsl(232, 23%, 7%)',
      },
      granite: {
        '100': 'hsl(0, 0%, 100%)',
        '098': 'hsl(20, 82%, 98%)',
        '096': 'hsl(60, 30%, 96%)',
        '084': 'hsl(0, 0%, 84%)',
        '063': 'hsl(210, 6%, 63%)',
        '060': 'hsl(180, 7%, 60%)',
        '045': 'hsl(196, 13%, 45%)',
        '040': 'hsl(194, 14%, 40%)',
        '00080': 'hsl(0, 0%, 0%, .8)',
        '000': 'hsl(0, 0%, 0%)',
      },
    },
    dark: {
      code: {
        background: 'hsl(231, 15%, 18%)',
        selection: 'hsl(232, 14%, 31%)',
        baseColor: 'hsl(60, 30%, 96%)',
        comment: 'hsl(225, 27%, 51%)',
        punctuation: 'hsl(60, 30%, 96%)',
        operator: 'hsl(326, 100%, 74%)',
        parameter: 'hsl(31, 100%, 71%)',
        property: 'hsl(265, 89%, 78%)',
        string: 'hsl(65, 92%, 76%)',
        function: 'hsl(135, 94%, 65%)',
        builtin: 'hsl(191, 97%, 77%)',
        keyword: 'hsl(326, 100%, 74%)',
        variable: 'hsl(265, 89%, 78%)',
      },
    },
    light: {
      code: {
        background: 'hsl(20, 82%, 98%)',
        selection: 'hsl(192, 81%, 14%)',
        baseColor: 'hsl(196, 13%, 45%)',
        comment: 'hsl(180, 7%, 60%)',
        punctuation: 'hsl(194, 14%, 40%)',
        operator: 'hsl(194, 14%, 40%)',
        parameter: 'hsl(18, 80%, 44%)',
        property: 'hsl(205, 69%, 49%)',
        string: 'hsl(175, 59%, 40%)',
        function: 'hsl(45, 100%, 35%)',
        builtin: 'hsl(205, 69%, 49%)',
        keyword: 'hsl(68, 100%, 30%)',
        variable: 'hsl(18, 80%, 44%)',
      },
    },
  },
  sizes: {
    base: {
      none: '0rem',
      px: '0.0625rem',
      xs: '0.125rem',
      sm: '0.25rem',
      md: '0.375rem',
      lg: '0.5rem',
      xl: '0.75rem',
      '2xl': '1rem',
      '3xl': '1.5rem',
      '4xl': '2rem',
      '5xl': '3rem',
      '6xl': '4rem',
      '7xl': '6rem',
      '8xl': '8rem',
      '9xl': '16rem',
      '10xl': '12rem',
      '11xl': '24rem',
      '12xl': '32rem',
      '13xl': '48rem',
      '14xl': '64rem',
      '15xl': '96rem',
      '16xl': '128rem',
    },
    border: {
      width: {
        none: '0rem',
        px: '0.0625rem',
        xs: '0.125rem',
        sm: '0.25rem',
        md: '0.375rem',
        lg: '0.5rem',
        xl: '0.75rem',
      },
      radius: {
        none: '0rem',
        px: '0.0625rem',
        xs: '0.125rem',
        sm: '0.25rem',
        md: '0.375rem',
        lg: '0.5rem',
        xl: '0.75rem',
        circle: '0.75rem',
        pill: '0.75rem',
      },
    },
    relative: {
      '0': '10',
      '1': '20',
      '2': '30',
      '3': '40',
      '4': '50',
      '5': '60',
      '6': '70',
      '7': '80',
      '8': '90',
      '9': '100',
    },
    percentage: {
      '10': '10%',
      '20': '20%',
      '30': '30%',
      '40': '40%',
      '50': '50%',
      '60': '60%',
      '70': '70%',
      '80': '80%',
      '90': '90%',
      '100': '100%',
    },
    screen: {
      width: {
        '10': '10vw',
        '20': '20vw',
        '30': '30vw',
        '40': '40vw',
        '50': '50vw',
        '60': '60vw',
        '70': '70vw',
        '80': '80vw',
        '90': '90vw',
        '100': '100vw',
      },
      height: {
        '10': '10vh',
        '20': '20vh',
        '30': '30vh',
        '40': '40vh',
        '50': '50vh',
        '60': '60vh',
        '70': '70vh',
        '80': '80vh',
        '90': '90vh',
        '100': '100vh',
      },
    },
    sizing: {
      none: '0rem',
      px: '0.0625rem',
      xs: '0.125rem',
      sm: '0.25rem',
      md: '0.375rem',
      lg: '0.5rem',
      xl: '0.75rem',
      '2xl': '1rem',
      '3xl': '1.5rem',
      '4xl': '2rem',
      '5xl': '3rem',
      '6xl': '4rem',
      '7xl': '6rem',
      '8xl': '8rem',
      '9xl': '16rem',
      '10xl': '12rem',
      '11xl': '24rem',
      '12xl': '32rem',
      '13xl': '48rem',
      '14xl': '64rem',
      '15xl': '96rem',
      '16xl': '128rem',
    },
    spacing: {
      none: '0rem',
      px: '0.0625rem',
      xs: '0.125rem',
      sm: '0.25rem',
      md: '0.375rem',
      lg: '0.5rem',
      xl: '0.75rem',
    },
  },
} as const;
