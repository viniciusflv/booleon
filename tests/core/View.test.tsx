import 'jest-styled-components';
import { View } from '../../packages/core/src';
import React from 'react';
import renderer from 'react-test-renderer';

const injectPseudoElement = (pseudo: string, component: any) => {
  component.props.className = component.props.className
    .split(/\s/)
    .map((classes: string) => ` ${classes}${pseudo}`)
    .join('');
  return component;
};

describe('Background', () => {
  test('bg_img', () => {
    const component = renderer
      .create(
        <View
          bc_red
          md__bc_red
          before__bc_red
          after__bc_red
          after__c_red
          hover__c_red
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
    expect(component).toHaveStyleRule('background-color', 'red');
    const after = injectPseudoElement(':hover', component);
    expect(after).toHaveStyleRule('color', 'red');
  });
});
