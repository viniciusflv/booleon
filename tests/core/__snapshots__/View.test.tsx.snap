// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Background bg_img 1`] = `
.c0 {
  background-color: red;
}

.c0:after {
  background-color: red;
  color: red;
}

.c0:before {
  background-color: red;
}

.c0:hover {
  color: red;
}

@media and screen (min-width:1024px) {
  .c0 {
    background-color: red;
  }
}

<div
  className="c0"
/>
`;
